src/vm/functions/sequences.rs

```diff
@@ -0,0 +1,177 @@
+use vm::errors::{CheckErrors, RuntimeErrorType, InterpreterResult as Result, check_argument_count};
+use vm::types::{Value, SequenceData, CharType, ListData, signatures::ListTypeData, TypeSignature::BoolType, TypeSignature};
+use vm::representations::{SymbolicExpression, SymbolicExpressionType};
+use vm::{LocalContext, Environment, CallableType, eval, apply, lookup_function};
+use std::convert::TryInto;
+use std::cmp;
+
+pub fn list_cons(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    let eval_tried: Result<Vec<Value>> =
+        args.iter().map(|x| eval(x, env, context)).collect();
+    let args = eval_tried?;
+
+    let mut arg_size = 0;
+    for a in args.iter() {
+        arg_size = arg_size.cost_overflow_add(a.size().into())?;
+    }
+
+    runtime_cost!(cost_functions::LIST_CONS, env, arg_size)?;
+
+    Value::list_from(args)
+}
+
+pub fn special_filter(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(2, args)?;
+
+    runtime_cost!(cost_functions::FILTER, env, 0)?;
+
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::ExpectedName)?;
+
+    let mut sequence = eval(&args[1], env, context)?;
+    let function = lookup_function(&function_name, env)?;
+
+    match sequence {
+        Value::Sequence(ref mut sequence_data) => {
+            sequence_data.filter(&mut |atom_value: SymbolicExpression| {
+                let argument = [atom_value];
+                let filter_eval = apply(&function, &argument, env, context)?;
+                if let Value::Bool(include) = filter_eval {
+                    return Ok(include);
+                } else {
+                    return Err(CheckErrors::TypeValueError(BoolType, filter_eval).into())
+                }
+            })?;
+        },
+        _ => return Err(CheckErrors::ExpectedSequence(TypeSignature::type_of(&sequence)).into())
+    };
+    Ok(sequence)
+}
+
+pub fn special_fold(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
+    check_argument_count(3, args)?;
+
+    runtime_cost!(cost_functions::FOLD, env, 0)?;
+
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::ExpectedName)?;
+
+    let function = lookup_function(&function_name, env)?;
+    let sequence = eval(&args[1], env, context)?;
+    let initial = eval(&args[2], env, context)?;
+
+    match sequence {
+        Value::Sequence(sequence_data) => {
+            sequence_data.atom_values()
+                .iter()
+                .try_fold(initial, |acc, x| {
+                    apply(&function, &[x.clone(), SymbolicExpression::atom_value(acc)], env, context)
+                })
+        },
+        _ => Err(CheckErrors::ExpectedSequence(TypeSignature::type_of(&sequence)).into())
+    }
+}
+
+pub fn special_map(args: &[SymbolicExpression], env: &mut Environment, context: &LocalContext) -> Result<Value> {
```

Comment:
Same comment here as for `special_fold` -- if `.atom_values()` was destructive, we could eliminate a copy. Also, I know we do this elsewhere, too, but we should use `.into_iter` instead of `.drain` when we're consuming the whole list.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1779/commits/27ff63ebd81f8d5566b87c04cbade46db81a2f51 / https://github.com/blockstack/stacks-blockchain/pull/1779/commits/613378c69228a6bde5213134ab4fcc671d993ea9.
Why preferring `into_iter()` over `drain`?

Reply:
`drain` leaves the mutated (and now empty) `Vec` object around until it goes out of scope and is dropped -- that means the memory allocation for it sticks around (even if it's empty, the Vec's memory allocation won't necessarily be "shrunk"). While that's not necessarily always a huge performance hit, I think the best practice is use `into_iter` whenever we can.